src/WebJobs.Script.WebHost/Extensions/ZipArchiveExtensions.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.IO.Compression;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Extensions
+{
+    public static class ZipArchiveExtensions
+    {
+        public static async Task AddDirectory(this ZipArchive zipArchive, DirectoryInfoBase directory, string directoryNameInArchive)
+        {
+            await InternalAddDirectory(zipArchive, directory, directoryNameInArchive);
+        }
+
+        private static async Task InternalAddDirectory(ZipArchive zipArchive, DirectoryInfoBase directory, string directoryNameInArchive, IList<ZipArchiveEntry> files = null)
+        {
+            bool any = false;
+            foreach (var info in directory.GetFileSystemInfos())
+            {
+                any = true;
+                if (info is DirectoryInfoBase subDirectoryInfo)
+                {
+                    string childName = ForwardSlashCombine(directoryNameInArchive, subDirectoryInfo.Name);
+                    await InternalAddDirectory(zipArchive, subDirectoryInfo, childName, files);
+                }
+                else
+                {
+                    var entry = await zipArchive.AddFile((FileInfoBase)info, directoryNameInArchive);
+                    files?.Add(entry);
+                }
+            }
+
+            if (!any)
+            {
+                // If the directory did not have any files or folders, add a entry for it
+                zipArchive.CreateEntry(EnsureTrailingSlash(directoryNameInArchive));
+            }
+        }
+
+        private static string ForwardSlashCombine(string part1, string part2)
+        {
+            return Path.Combine(part1, part2).Replace('\\', '/');
+        }
+
+        public static Task<ZipArchiveEntry> AddFile(this ZipArchive zipArchive, string filePath, string directoryNameInArchive = "")
+        {
+            var fileInfo = FileUtility.FileInfoFromFileName(filePath);
+            return zipArchive.AddFile(fileInfo, directoryNameInArchive);
+        }
+
+        public static async Task<ZipArchiveEntry> AddFile(this ZipArchive zipArchive, FileInfoBase file, string directoryNameInArchive)
+        {
+            Stream fileStream = null;
+            try
+            {
+                fileStream = file.OpenRead();
+            }
+            catch (Exception)
+            {
+                // tolerate if file in use.
```

Comment:
What will the behavior below be if the file doesn't exist? And if it is in use, how will this behave when trying to copy below?

Reply:
if the file doesn't exist, `file.OpenRead()` will throw, and this method will just return null.
If the file is in an exclusive read, then this will also throw, and we won't add that locked file to the zip. Note that it isn't for any locked file. I'm calling `.OpenRead()` which just tried the minimum for reading the file. If we can't read the file, this will skip it. Do you think it should fail instead?

Reply:
That's what I was wondering. I would be inclined to say failing is safer, but I'm fine with what you decide to do.

Reply:
I'll change it.